src/ADAL.PCL/TokenCache.cs

```diff
@@ -470,15 +547,25 @@ private void UpdateCachedMrrtRefreshTokens(AuthenticationResultEx result, string
         /// authority value that this AuthorizationContext was created with.  In every case passing
         /// null results in a wildcard evaluation.
         /// </summary>
-        private List<KeyValuePair<TokenCacheKey, AuthenticationResultEx>> QueryCache(string authority, string clientId, TokenSubjectType subjectType, string uniqueId, string displayableId)
+        private List<KeyValuePair<TokenCacheKey, AuthenticationResultEx>> QueryCache(string authority, string clientId,
+            TokenSubjectType subjectType, string uniqueId, string displayableId)
         {
-            return this.tokenCacheDictionary.Where(
+            cacheLock.EnterReadLock();
```

Comment:
Its a write lock and it can recursively enter the other two modes.The documentation mentions of this property